cubids/utils.py

```diff
@@ -1034,3 +1034,91 @@ def assign_variants(summary, rename_cols):
                 summary.at[row, col] = ""
 
     return summary
+
+
+def collect_file_collections(layout, base_file):
+    """Build a list of files in a file collection for a given base file.
+
+    Parameters
+    ----------
+    layout : BIDSLayout
+        The BIDSLayout object.
+    base_file : str
+        The base file to collect file collections for.
+
+    Returns
+    -------
+    files : list of BIDSFile
+        A list of files in the file collection for the given base file.
+    out_metadata : dict
+        A dictionary of metadata for the file collection, to be added to each file's metadata.
+    """
+    from bids.layout import Query
+
+    file_collection_entities = {
+        "echo": "EchoTime",
+        "part": None,
+        "mt": "MTState",
+        "inv": "InversionTime",
+        "flip": "FlipAngle",
+    }
+
+    base_file = layout.get_file(base_file)
+    fc_query = {ent: [Query.ANY, Query.NONE] for ent in file_collection_entities.keys()}
+    query = base_file.get_entities()
+    query = {**query, **fc_query}
+    files = layout.get(**query)
+
+    if len(files) <= 1:
+        return files, {}
+
+    # Get list of entities present in any of the files
+    collected_entities = [list(f.get_entities().keys()) for f in files]
+    # Flatten the list
+    collected_entities = [item for sublist in collected_entities for item in sublist]
+    # Remove duplicates
+    collected_entities = sorted(set(collected_entities))
+
+    out_metadata = {}
+    # Add metadata field with BIDS URIs to all files in file collection
+    out_metadata["FileCollection"] = [get_bidsuri(f.path, layout.root) for f in files]
+
+    files_metadata = [f.get_metadata() for f in files]
```

Comment:
I could swap this out with `sidecar = utils.img_to_new_ext(str(collection_path), ".json")` but that will ignore the inheritance principle. On the other hand, indexing metadata might be the biggest time cost for initializing the BIDSLayout object on large datasets, so it might not be worth it to account for inherited metadata.

Reply:
I dropped the metadata-indexing since we mostly use CuBIDS on large datasets.